kernel/uGraph.ml

```diff
@@ -786,6 +786,42 @@ let constraints_of_universes g =
   let csts = UMap.fold constraints_of g.entries Constraint.empty in
   csts, UF.partition uf
 
+(* domain g.entries = kept + removed *)
+let constraints_for ~kept g =
+  (* rmap: partial map from canonical universes to kept universes *)
+  let rmap, csts = LSet.fold (fun u (rmap,csts) ->
+      let arcu = repr g u in
+      if LSet.mem arcu.univ kept then
+        LMap.add arcu.univ arcu.univ rmap, enforce_eq_level u arcu.univ csts
+      else
+        match LMap.find arcu.univ rmap with
+        | v -> rmap, enforce_eq_level u v csts
+        | exception Not_found -> LMap.add arcu.univ u rmap, csts)
```

Comment:
Here I'm a bit worried: rmap is in the process of being built, so we depend on the fold order somehow?

Otherwise, it looks to me that indeed, constraints_for will for sure not return constraints mentioning anything else than the kept universes.

Reply:
>Here I'm a bit worried: rmap is in the process of being built, so we depend on the fold order somehow?

The fold order impacts which kept universe is canonical when the ugraph considers a removed universe to be canonical.

Reply:
@SkySkimmer could you confirm or infirm my worry?

Reply:
What is the worry exactly? ie what can go wrong?

Reply:
Good, thanks for the explanation.